### PR TITLE
[DDO-1728] Improve metric name construction to prepare for Prometheus

### PR DESCRIPTION
**Pull request by jack-r-warren**
**Edit:** I've encountered more issues here, so putting this on pause.

---

**Short version:** rather than an InstrumentationPath being a `NonEmptyList[String]`, it is now [a more complex type](https://github.com/broadinstitute/cromwell/blob/DDO-1728-handle-metric-variance/services/src/main/scala/cromwell/services/instrumentation/CromwellInstrumentation.scala#L48). Everything else is refactoring to match that change.

**Long version:**

There's basically two competing models for metric names--"lots of metrics, no labels" and "smaller a number of metrics, with labels for variation". Statsd uses the first, and Prometheus (and Stackdriver too, ideally) use the second.

By way of example:

```
api.response.count.200
```

versus

```
api_response_count{code="200"}
```

This boils down to how the different systems query metrics. In a StatsD world, you can have a query like `api.response.count.*`, but that's not possible for Prometheus/Stackdriver--the actual metric name needs to be fully static, and you conceptually do `api_response_count{code="*"}`.

Cromwell metric names right now are this:

```scala
type InstrumentationPath = NonEmptyList[String]

CromwellBucket(prefix: List[String], path:  InstrumentationPath)
```

Metric names are guaranteed to not be empty by the path. The path is what is actually assembled in code and passed around, and prefix (or an empty list) is added at the outer edge--most of Cromwell treats the `NonEmptyList[String]` as the metric name. When the metrics are actually sent to StatsD, the strings are joined with periods and sent off.

I spent several days trying to figure out a way to reliably parse labels out from these names, and I couldn't figure it out. There's two key pieces of info known at the time the `NonEmptyList[String]` is assembled that are not recorded:
- When a particular string being added to the list is expected to vary frequently (and thus needs to be a label)
- What that string means (labels are key-value pairs and we're usually missing the key)

> This parsing-based approach actually did work with Rawls, by doing the parsing outside Rawls's codebase, not inside it. That worked for Rawls because Rawls has relatively few actually unique metrics, and the metrics that it does have are extremely stable--no one is going to change them--so we literally hardcode the exact glob matching and regex patterns. With the number of metric Cromwell has, that raw hardcoding approach would get unmaintainable very, very quickly.

My idea was to instead have metric names effectively be the following:

```scala
type InstrumentationPath = NonEmptyList[String | (String, String)]

CromwellBucket(prefix: List[String], path:  InstrumentationPath)
```

Parts of a metric name expected to vary frequently could be stored as a key-value pair in the same list, and StatsD and Prometheus could simply handle it differently: StatsD could ignore the key and use the value to get the exact metric name as before, while Prometheus could assemble the name from just the strings and pass the key-value pairs as the labels.

Unfortunately, union types are a Scala 3 thing. In Scala 2 we get `Either`, a manually-boxed type. :((((

So I did this:

```scala
private type LowVariantPart = String
private type HighVariantPart = (String, String)
private type InternalPath = NonEmptyList[Either[LowVariantPart, HighVariantPart]]

implicit class InstrumentationPath (val internalPath: InternalPath) { ... }
object InstrumentationPath { ... }

CromwellBucket(prefix: List[String], path:  InstrumentationPath)
```

The implicit class provides a similar conceptual interface to the NonEmptyList, except it handles the necessary boxing (it also the includes some convenience methods that were already on a `implicit class EnhancedStatsDPath(val path: InstrumentationPath) extends AnyVal`). The object effectively provides constructors to handle the now-internal NonEmptyList invariant.

What does all this mean?

Some screenshots of the sorts of changes this causes:
1 -
Screen Shot 2022-02-16 at 4 30 35 PM
2 -
Screen Shot 2022-02-16 at 4 31 20 PM
3 -
Screen Shot 2022-02-16 at 4 31 34 PM
4 -
Screen Shot 2022-02-16 at 4 36 36 PM

There's a file of passing tests for the new InstrumentationPath class [here](https://github.com/broadinstitute/cromwell/blob/DDO-1728-handle-metric-variance/services/src/test/scala/cromwell/services/instrumentation/CromwellInstrumentationSpec.scala).

The bottom line is that this keeps metric names going to StatsD and Stackdriver the same while allowing Prometheus to create different names/labels with the same info.

> (FWIW, Stackdriver could probably benefit from labels as well, because I don't know that it'd be possible to create good graph queries against Cromwell's current Stackdriver stats. For here, though, I'm trying to get perfect backwards compatibility)

### CODE REVIEW

**Comment by jgainerdewar** [comment]
Hey @jack-r-warren, thanks for your patience! I think this approach overall looks good, but I'm having some trouble understanding how more complex `InstrumentationPath` objects translate to Prometheus metric identifiers. How does the below `IP` translate to the `api_response_count{code="200"}` type notation you used above?

```
InstrumentationPath
    .withParts("a")
    .withHighVariantPart("label-b" -> "b")
    .withParts("c", "d")
    .withHighVariantPart("label-e", "e")
    .withParts(List("f", "g"))
```

Is it `a.c.d.f.g{label-b="b", label-e="e"}`? Or is the original ordering information preserved somehow?

**Comment by jack-r-warren** [comment]
@jgainerdewar Your guess was correct--Prometheus/Stackdriver labels have no concept of order, so that information is lost when the InstrumentationPath is used to derive that sort of name. The StatsD form effectively looses the additional "what's a high variant part and what isn't" info while the Prometheus form looses the ordering of those parts. This is normal in Prometheus-land and all the query systems are built on that assumption. 

There'll definitely be tests for the Prometheus code when it's added to clearly demonstrate that--this is sorta a weird one because I want to lay the groundwork without actually adding full Prometheus support yet. [This test](https://github.com/broadinstitute/cromwell/pull/6681/files#diff-65d7248a0c8799434124fe6d53023d0d8ac3492d640bbe4510801d2981e0903fR42) and a few others check that the groundwork is working, it's just a bit of a weird refactoring until the Prometheus code takes advantage of it.

**Comment by jgainerdewar** [comment]
Okay, makes sense! I think this approach is an elegant solution to the problem. As far as I'm concerned you're good to go forward with finishing this change.

**Review by mcovarr — COMMENTED** [review]
Just a couple of high level comments. I agree with Janet that this seems like a nice way of handling the differences between the metrics models. Thanks again for taking this on! 👍

File: CromIAM/src/main/scala/cromiam/instrumentation/CromIamInstrumentation.scala [hunk 1]
```diff
@@ -18,39 +17,38 @@ object CromIamInstrumentation {
...
 trait CromIamInstrumentation extends CromwellInstrumentation {
 
-  val samPrefix: NonEmptyList[String] = NonEmptyList.one("sam")
-  val getWhitelistPrefix = NonEmptyList.one("get-whitelist")
-  val userCollectionPrefix = NonEmptyList.one("user-collection")
-  val authCollectionPrefix = NonEmptyList.one("auth-collection")
-  val registerCollectionPrefix = NonEmptyList.one("register-collection")
-  val rootWfIdPrefix = NonEmptyList.one("root-workflow-id")
-  val wfCollectionPrefix = NonEmptyList.one("workflow-collection")
+  val samPrefix = InstrumentationPath.withParts("sam")
```
[the remark this answers is not in the document]
Comment (jack-r-warren) [reply CromIAM/src/main/scala/cromiam/instrumentation/CromIamInstrumentation.scala]: Fair enough--added types to the public fields/methods this PR touches

File: services/src/main/scala/cromwell/services/instrumentation/CromwellInstrumentation.scala [hunk 2]
```diff
@@ -21,19 +20,73 @@ object CromwellInstrumentation {
...
+   * This distinction exists because instrumentation services meant to handle multiple time series per metric name
+   * (like Prometheus) are poorly built to store numerous metric names or query across them. The more complex
+   * [[getPathAndLabels]] output allows the backends for those instrumentation services to "play nice" with their
+   * associated ecosystems.
+   *
+   * @param internalPath the runtime representation of this class, where the 'left' is a simple, always-present
+   *                     low-variant name part, and the 'right' is a key-value pair for a frequently-changing
+   *                     high-variant name part
+   */
+  implicit class InstrumentationPath (val internalPath: NonEmptyList[Either[String, (String, String)]]) extends AnyVal {
```
[the remark this answers is not in the document]
Comment (jack-r-warren) [reply services/src/main/scala/cromwell/services/instrumentation/CromwellInstrumentation.scala]: Makes sense, I added some internal types to that class and used those for documentation. I figured I wouldn't expose them outside the class since I think using them as method inputs or something would just be unnecessary indirection

**Comment by jack-r-warren** [comment]
(Moving this back to draft while I tweak something that came up while adding Prometheus support on a local branch--I think it makes sense to do in this PR rather than the next)
